File: branches/1.1_before_major_refactor_2.0/csharp/core/command/ICommand.cs
using System;
using System.Collections.Generic;
using System.Text;
using urakawa.xuk;
using urakawa.events;
using urakawa.events.undo;
using ExecutedEventArgs=urakawa.events.command.ExecutedEventArgs;
using UnExecutedEventArgs=urakawa.events.command.UnExecutedEventArgs;

namespace urakawa.command
{
    /// <summary>
    /// Classes realizing this interface must store the state of the object(s) affected by the command
    /// execution (including exception/redo). Implementations may choose various techniques suitable in terms
    /// of performance and memory usage (storage of the transition or the full object snapshot.)
    /// </summary>
    public interface ICommand : IWithPresentation, IXukAble, IChangeNotifier, IAction
    {
        /// <summary>
        /// Event fired after the <see cref="ICommand"/> has been executed
        /// </summary>
        event EventHandler<ExecutedEventArgs> executed;

        /// <summary>
        /// Event fired after the <see cref="ICommand"/> has been un-executed
        /// </summary>
        event EventHandler<UnExecutedEventArgs> unExecuted;

        /// <summary>
        /// Execute the reverse command.
        /// </summary>
        /// <exception cref="urakawa.exception.CannotUndoException">Thrown when the command cannot be reversed.</exception>
        void unExecute();

        /// <summary>
        /// True if the command is reversible.
        /// </summary>
        bool canUnExecute();

        /// <summary>
        /// Gets a list of the <see cref="media.data.MediaData"/> used by the Command
        /// </summary>
        /// <returns></returns>
        List<media.data.MediaData> getListOfUsedMediaData();
    }
}

File: trunk/java/core/src/org/daisy/urakawa/command/ICommand.java
package org.daisy.urakawa.command;

import java.util.List;

import org.daisy.urakawa.IWithPresentation;
import org.daisy.urakawa.events.IEventHandler;
import org.daisy.urakawa.events.command.CommandEvent;
import org.daisy.urakawa.media.data.IMediaData;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * <p>
 * Classes realizing this interface must store the state of the object(s)
 * affected by the command execution.
 * </p>
 * 
 * @depend - Aggregation 1 org.daisy.urakawa.IPresentation
 * @depend - Event - org.daisy.urakawa.event.undo.CommandExecutedEvent
 * @depend - Event - org.daisy.urakawa.event.undo.CommandUnExecutedEvent
 */
public interface ICommand extends IAction, IXukAble, IWithPresentation,
        IEventHandler<CommandEvent>
{
    /**
     * <p>
     * Returns a list of IMediaData objects that are in use by this command.
     * </p>
     * 
     * @return a non-null, possibly empty, list of IMedia objects
     */
    public List<IMediaData> getListOfUsedMediaData();

    /**
     * <p>
     * executes the reverse ICommand
     * </p>
     * 
     * @tagvalue Events "CommandUnExecuted"
     * 
     * @throws CommandCannotUnExecuteException when the ICommand cannot be
     *         un-executed
     */
    public void unExecute() throws CommandCannotUnExecuteException;

    /**
     * <p>
     * Tests whether this command can be un-executed.
     * </p>
     * 
     * @return true if this command can be un-executed.
     */
    public boolean canUnExecute();
}
